src/merkle/tiered_smt/mod.rs

```diff
@@ -132,51 +89,81 @@ impl TieredSmt {
-    /// Tree defined by the same root as this tree.
-    pub fn prove(&self, key: RpoDigest) -> TieredSmtProof {
-        let (path, index, leaf_exists) = self.nodes.get_proof(&key);
+    /// The iterator order is unspecified.
+    pub fn inner_nodes(&self) -> impl Iterator<Item = InnerNodeInfo> + '_ {
+        self.nodes.inner_nodes()
+    }
+}
 
-        let entries = if index.depth() == Self::MAX_DEPTH {
-            match self.values.get_all(index.value()) {
-                Some(entries) => entries,
-                None => vec![(key, Self::EMPTY_VALUE)],
-            }
-        } else if leaf_exists {
-            let entry =
-                self.values.get_first(index_to_prefix(&index)).expect("leaf entry not found");
-            debug_assert_eq!(entry.0, key);
-            vec![*entry]
-        } else {
-            vec![(key, Self::EMPTY_VALUE)]
-        };
+impl<T> TieredSmt<T>
+where
+    T: Copy,
+{
+    /// Returns an iterator over bottom leaves (i.e., depth = 64) of this [TieredSmt].
+    ///
+    /// Each yielded item consists of the hash of the leaf and its contents, where contents is
+    /// a vector containing key-value pairs of entries storied in this leaf.
+    ///
+    /// The iterator order is unspecified.
+    pub fn bottom_leaves(&self) -> impl Iterator<Item = (RpoDigest, Vec<(RpoDigest, T)>)> + '_ {
+        self.nodes.bottom_leaves().map(|(&prefix, node)| {
+            let values = self.values.get_all(prefix).expect("bottom leaf not found");
+            (*node, values)
+        })
+    }
 
-        TieredSmtProof::new(path, entries).expect("Bug detected, TSMT produced invalid proof")
+    /// Returns an iterator over upper leaves (i.e., depth = 16, 32, or 48) for this [TieredSmt]
+    /// where each yielded item is a (node, key, value) tuple.
+    ///
+    /// The iterator order is unspecified.
+    pub fn upper_leaves(&self) -> impl Iterator<Item = (RpoDigest, RpoDigest, T)> + '_ {
+        self.nodes.upper_leaves().map(|(index, node)| {
+            let key_prefix = index_to_prefix(index);
+            let (key, value) = self.values.get_first(key_prefix).expect("upper leaf not found");
+            debug_assert_eq!(*index, LeafNodeIndex::from_key(key, index.depth()).into());
+            (*node, *key, *value)
+        })
     }
+}
 
-    // STATE MUTATORS
-    // --------------------------------------------------------------------------------------------
+impl<T> TieredSmt<T>
+where
+    T: Default + Copy,
+{
+    /// Returns the value associated with the specified key.
+    ///
+    /// If nothing was inserted into this tree for the specified key, [ZERO; 4] is returned.
+    pub fn get_value(&self, key: RpoDigest) -> T {
+        match self.values.get(&key) {
+            Some(value) => *value,
+            None => T::default(),
+        }
+    }
+}
 
+impl<T> TieredSmt<T>
+where
+    T: Default + Copy + PartialEq + Debug + Into<RpoDigest>,
```

Comment:
I used `Into<RpoDigest>` to constraint the objects, this is used to compute the hash of the leaves.

This may be a good or a bad thing. The previous version used `Word`, which meant objects that can be represented with a single Word of data. This now says objects that can be hashed to a RpoDigest. This conversion is more general, but may not be what we want, for example, an `Asset` isn't really a hash value, but data that fits in a Word.